src/main/java/org/roaringbitmap/RoaringIterator.java

```diff
@@ -0,0 +1,138 @@
+package org.roaringbitmap;
+
+import java.util.NoSuchElementException;
+
+class RoaringIterator implements IntIterator {
+  final private RoaringBitmap parent;
+  private ArrayContainer ac;
+  private RunContainer rc;
+  private int containerIndex;
+  private short typecode = TypeCode.NONE;
+  private int hb;
+  private int le = 0;
+  private int maxlength;
+  private int base;
+  private long w;
+  private int pos;
+  private long[] bitmap;
+
+  RoaringIterator(RoaringBitmap parent) {
+    this.parent = parent;
+    this.hb = 0;
+    this.containerIndex = 0;
+    nextContainer();
+  }
+
+  public boolean hasNext() {
+    return containerIndex <= parent.highLowContainer.size;
+  }
+
+  public int next() {
+    int answer;
+    switch (typecode) {
+      case TypeCode.BITSET_CONTAINER:
+        long t = w & -w;
```

Comment:
Could this not be done polymorphically on Container? It seems to be that the type codes are used in several places in CRoaring (where indeed polymorphism _is_ used in the equivalent java features) because there is just no runtime override resolution in C.